src/x/parsers/metric_parsers.go

```diff
@@ -0,0 +1,69 @@
+package parsers
+
+import (
+	"github.com/m3db/m3/src/metrics/metric/id"
+	"github.com/m3db/m3/src/x/pool"
+	"github.com/m3db/m3/src/x/serialize"
+	"bytes"
+	"github.com/m3db/m3/src/x/ident"
+	"github.com/m3db/m3/src/x/checked"
+)
+
+func getMetricTagsIterator() serialize.MetricTagsIterator {
+	// todo: creating a tag decoder in an ad-hoc manner, replace with our own decoder without the need of a pool
+	poolOpts := pool.NewObjectPoolOptions().SetSize(1)
+	size := 1
+	tagDecoderPool := serialize.NewTagDecoderPool(
```

Comment:
the `shardFn` is being called for every time series, it is a bit resource intensive if every time we need to create new `TagDecodePool`, the test showed aggregator cpu jumps in dev, I'd suggest we shall not directly using the pool but just making the iterator a `static method` with onepass:

<img width="726" alt="le" src="https://github.com/databricks/m3/assets/96499497/283cf38c-3363-4ad2-8d1b-602507af5513">

Reply:
+1, and even further we don't need to decode it each time and maintain a mapping {RawID : {containsLe : Boolean, newRawID:idWihtoutLE}}. 

Then you don't have to construct this decoder and make the copy each time for each RawID and iterate all the labels. I assume rawID size is limited, I don't think this will cause too much memory consuming, but need to verify.

@rishangirish can you also verify the memory utilization and put cpu/memory into pr description?

Reply:
> Does this original vs pre-aggregated mismatch align with this deployment in the graph? Looks like it matches well before
> 
> ![Screenshot 2023-05-12 at 12 53 05 PM](https://user-images.githubusercontent.com/5896536/238066688-4448ee22-1568-4481-865b-dd1f678b6402.png)

Correct, that does appear to be a slight error margin for the agg vs. original metric and i think this has to do with the delay that is introduced. but if we expand the time window, we see that the two metrics are the same: 
<img width="819" alt="image" src="https://github.com/databricks/m3/assets/85959671/7d3c4cc3-d344-4e36-b1c3-3c04d54335ea">